File: source/nndeploy/device/type.cc

#include "nndeploy/device/type.h"

namespace nndeploy {
namespace device {

BufferDesc::BufferDesc() {}
BufferDesc::BufferDesc(size_t size) { size_.emplace_back(size); }
BufferDesc::BufferDesc(size_t *size, size_t len) {
  for (int i = 0; i < len; ++i) {
    size_.emplace_back(size[i]);
  }
}
BufferDesc::BufferDesc(const base::SizeVector &size,
                       const base::IntVector &config)
    : size_(size), config_(config) {}
BufferDesc::BufferDesc(size_t *size, size_t len, const base::IntVector &config)
    : config_(config) {
  for (int i = 0; i < len; ++i) {
    size_.emplace_back(size[i]);
  }
}

BufferDesc::BufferDesc(const BufferDesc &desc) {
  if (this == &desc) {
    return;
  }
  size_ = desc.size_;
  config_ = desc.config_;
}
BufferDesc &BufferDesc::operator=(const BufferDesc &desc) {
  if (this == &desc) {
    return *this;
  }
  size_ = desc.size_;
  config_ = desc.config_;
  return *this;
}
BufferDesc &BufferDesc::operator=(size_t size) {
  size_.emplace_back(size);
  return *this;
}

BufferDesc::BufferDesc(BufferDesc &&desc) noexcept {
  if (this == &desc) {
    return;
  }
  size_ = std::move(desc.size_);
  config_ = std::move(desc.config_);
}
BufferDesc &BufferDesc::operator=(BufferDesc &&desc) noexcept {
  if (this == &desc) {
    return *this;
  }
  size_ = std::move(desc.size_);
  config_ = std::move(desc.config_);
  return *this;
}

BufferDesc::~BufferDesc(){};

bool BufferDesc::isSameConfig(const BufferDesc &desc) const {
  if (config_.size() != desc.config_.size()) {
    return false;
  }
  for (int i = 0; i < config_.size(); ++i) {
    if (config_[i] != desc.config_[i]) {
      return false;
    }
  }
  return true;
}
bool BufferDesc::isSameDim(const BufferDesc &desc) const {
  if (size_.size() == desc.size_.size()) {
    return true;
  } else {
    return false;
  }
}
bool BufferDesc::is1D() const { return size_.size() == 1; }

bool BufferDesc::operator>=(const BufferDesc &other) const {
  bool flag = true;
  for (int i = 0; i < size_.size(); ++i) {
    if (size_[i] < other.size_[i]) {
      flag = false;
      break;
    }
  }
  return flag;
}
bool BufferDesc::operator==(const BufferDesc &other) const {
  bool flag = true;
  for (int i = 0; i < size_.size(); ++i) {
    if (size_[i] != other.size_[i]) {
      flag = false;
      break;
    }
  }
  return flag;
}
bool BufferDesc::operator!=(const BufferDesc &other) const {
  return !(*this == other);
}

// TensorDesc
TensorDesc::TensorDesc(){};

TensorDesc::TensorDesc(base::DataType data_type, base::DataFormat format,
                       const base::IntVector &shape)
    : data_type_(data_type), data_format_(format), shape_(shape){};

TensorDesc::TensorDesc(base::DataType data_type, base::DataFormat format,
                       const base::IntVector &shape,
                       const base::SizeVector &stride)
    : data_type_(data_type),
      data_format_(format),
      shape_(shape),
      stride_(stride){};

TensorDesc::TensorDesc(const TensorDesc &desc) {
  if (this == &desc) {
    return;
  }
  data_type_ = desc.data_type_;
  data_format_ = desc.data_format_;
  shape_ = desc.shape_;
  stride_ = desc.stride_;
}
TensorDesc &TensorDesc::operator=(const TensorDesc &desc) {
  if (this == &desc) {
    return *this;
  }
  data_type_ = desc.data_type_;
  data_format_ = desc.data_format_;
  shape_ = desc.shape_;
  stride_ = desc.stride_;
}

TensorDesc::TensorDesc(TensorDesc &&desc) noexcept {
  if (this == &desc) {
    return;
  }
  data_type_ = desc.data_type_;
  data_format_ = desc.data_format_;
  shape_ = std::move(desc.shape_);
  stride_ = std::move(desc.stride_);
}
TensorDesc &TensorDesc::operator=(TensorDesc &&desc) noexcept {
  if (this == &desc) {
    return *this;
  }
  data_type_ = desc.data_type_;
  data_format_ = desc.data_format_;
  shape_ = std::move(desc.shape_);
  stride_ = std::move(desc.stride_);
}

TensorDesc::~TensorDesc(){};

bool TensorDesc::operator==(const TensorDesc &other) const {
  bool flag0 = false;
  if (shape_.size() == other.shape_.size()) {
    flag0 = true;
    for (int i = 0; i < shape_.size(); ++i) {
      if (shape_[i] != other.shape_[i]) {
        flag0 = false;
        break;
      }
    }
  }
  bool flag1 = false;
  if (stride_.size() == other.stride_.size()) {
    flag1 = true;
    for (int i = 0; i < stride_.size(); ++i) {
      if (stride_[i] != other.stride_[i]) {
        flag1 = false;
        break;
      }
    }
  }
  bool flag2 = data_type_ == other.data_type_;
  bool flag3 = data_format_ == other.data_format_;
  return flag0 && flag1 && flag2 && flag3;
}
bool TensorDesc::operator!=(const TensorDesc &other) const {
  return !(*this == other);
}

}  // namespace device
}  // namespace nndeploy

File: python/nndeploy/device/type.py

import nndeploy._nndeploy_internal as _C

import nndeploy.base


class BufferDesc(_C.device.BufferDesc):
    def __init__(self, *args, **kwargs):
        """
        Constructs a BufferDesc object.

        The constructor can be called in the following ways:
        1. BufferDesc(): Constructs an empty BufferDesc.
        2. BufferDesc(size): Constructs a BufferDesc with size.
        3. BufferDesc(size_ptr, size): Constructs a BufferDesc from a size_t array of size.
        4. BufferDesc(size_vector): Constructs a BufferDesc from a base::SizeVector.
        5. BufferDesc(size, int_vector): Constructs a BufferDesc from a size and a base::IntVector.
        6. BufferDesc(size_vector, int_vector): Constructs a BufferDesc from a base::SizeVector and a base::IntVector.
        7. BufferDesc(size_ptr, size, int_vector): Constructs a BufferDesc from a size_t array of size and a base::IntVector.
        """
        super().__init__(*args, **kwargs)
    
    def __eq__(self, other):
        return super().__eq__(other)
    
    def __ne__(self, other):
        return super().__ne__(other)
    
    def __ge__(self, other):
        return super().__ge__(other)
    
    def get_size(self):
        return super().get_size()
    
    def get_size_vector(self):
        return super().get_size_vector()
    
    def get_real_size(self):
        return super().get_real_size()
    
    def get_real_size_vector(self):
        return super().get_real_size_vector()
    
    def get_config(self):
        return super().get_config()
    
    def is_same_config(self, other):
        return super().is_same_config(other)
    
    def is_same_dim(self, other):
        return super().is_same_dim(other)
    
    def is_1d(self):
        return super().is_1d()
    
    def print(self, stream):
        super().print(stream)
    
    def just_modify(self, *args):
        return super().just_modify(*args)
    
    def clear(self):
        super().clear()
    
    def __str__(self):
        return super().__str__()

class TensorDesc(_C.device.TensorDesc):
    def __init__(self, *args, **kwargs):
        """
        Constructs a TensorDesc object.

        The constructor can be called in the following ways:
        1. TensorDesc(): Constructs an empty TensorDesc object.
        2. TensorDesc(data_type, format, shape): Constructs a TensorDesc object from data type, data format and shape.
        3. TensorDesc(data_type, format, shape, stride): Constructs a TensorDesc object from data type, data format, shape and stride.
        4. TensorDesc(desc): Constructs a new TensorDesc object from another TensorDesc object.
        """
        super().__init__(*args, **kwargs)

    def __eq__(self, other):
        return super().__eq__(other)

    def __ne__(self, other):  
        return super().__ne__(other)

    def print(self, stream):
        super().print(stream)

    @property
    def data_type(self):
        return self.data_type_

    @data_type.setter
    def data_type(self, value):
        self.data_type_ = value

    @property
    def data_format(self):
        return self.data_format_

    @data_format.setter 
    def data_format(self, value):
        self.data_format_ = value

    @property
    def shape(self):
        return self.shape_

    @shape.setter
    def shape(self, value):  
        self.shape_ = value

    @property
    def stride(self):
        return self.stride_

    @stride.setter
    def stride(self, value):
        self.stride_ = value

    def __str__(self):
        return super().__str__()
  
    

    
    
